Fix chunk handling for entropy scanner

rbailey-godaddy opened the pull request:
It's not actually true that *all* chunks are diff output; local folder scans have just raw content (prefixed with filename). To fix this:

* Add a new "is_diff" property to the Chunk type. We do this instead of adding this to the existing metadata property so that we do not perturb user-visible output (where we blindly render every member).
* Fix all Chunk generators to set .is_diff properly.
* Adjust scan_entropy() so that it removes the first character from each line of a chunk when .is_diff is True, and leaves it alone otherwise. The first character is preserved in expectation that we might use it for some backwards-compatibility kludge (but right now it just pisses off pylint because it's never consumed).

Unit tests updated, mostly to just add the missing constructor parameter. The local folder scan tests now pass again.

Conundrums left for another day:

* Technically, scan_regex() should be doing the same thing, but this also would introduce a compatiblity break and it's more effort because we presently don't do any chunk interpretation. It's probably better to just leave it as-is.
* I notice that bizarrely one of the pre-commit unit tests fails if you have staged files present. It works fine as long as you don't have any such files. Based on what it claims to be doing, this behavior shouldn't be occurring in the first place.

rscottbailey committed:
Fix chunk handling for entropy scanner

It's not actually true that *all* chunks are diff output; local folder
scans have just raw content (prefixed with filename). To fix this:

* Add a new "is_diff" property to the Chunk type. We do this instead of
  adding this to the existing metadata property so that we do not
  perturb user-visible output (where we blindly render every member).
* Fix all Chunk generators to set .is_diff properly.
* Adjust scan_entropy() so that it removes the first character from each
  line of a chunk when .is_diff is True, and leaves it alone otherwise.
  The first character is preserved in expectation that we might use it
  for some backwards-compatibility kludge (but right now it just pisses
  off pylint because it's never consumed).

Unit tests updated, mostly to just add the missing constructor
parameter. The local folder scan tests now pass again.

Conundrums left for another day:

* Technically, scan_regex() should be doing the same thing, but this
  also would introduce a compatiblity break and it's more effort because
  we presently don't do any chunk interpretation. It's probably better
  to just leave it as-is.
* I notice that bizarrely one of the pre-commit unit tests fails if you
  have staged files present. It works fine as long as you don't have any
  such files. Based on what it claims to be doing, this behavior
  shouldn't be occurring in the first place.

rscottbailey committed:
Support old entropy issue signatures

I think. This code isn't tested yet. Overhead should be relatively small
(aka manageable) because we go to the effort of checking only if:
* There is an entropy finding
* It isn't excluded
* It appears at the beginning of a line

In that situation, we retry the exclusion based on the string that
tartufo used to use. If THAT gets excluded, then we:
* Do not generate an issue
* We do create issues (temporarily) for both the old and new strings
* We report the old and new signatures and tell users to update their
  configuration

sushantmimani commented on tartufo/scanner.py:
```diff
@@ -591,40 +591,82 @@ def scan_entropy(
                 for string in util.find_strings_by_regex(word, BASE64_REGEX):
                     yield from self.evaluate_entropy_string(
-                        chunk, line, string, self.b64_entropy_limit
+                        chunk, line, string, self.b64_entropy_limit, extra_char
```
Shouldn't this be `anazlye`? And same for the next one?
```suggestion
                        chunk, analyze, string, self.b64_entropy_limit, extra_char
```

rscottbailey committed:
Fix typos

All instances of line should have been replaced by analyze.
Spotted by @sushantmimani
